src/Property.php

```diff
@@ -143,6 +143,41 @@ protected function cast($value)
         return new $castTo($value);
     }
 
+    protected function castArray(array $values)
+    {
+        foreach ($values as $value) {
+            if (! is_array($value)) {
+                return $this->cast($values);
+            }
+        }
+
+        $castTo = null;
+        $flatTypes = str_replace('[]', '', $this->types);
```

Comment:
Perhaps it'd be better to parse the flatTypes in `Property:: resolveTypeDefinition`, so that we don't have to do all this "setup work" while casting.

Reply:
This seems like a better place to put the responsibility of the type definition, yes.